designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+
+API changes create a user migration burden that should be weighed against the benefits of the breaking changes. Batching breaking changes into a single version bump **helps to minimize this burden**. The v1alpha5 API has seen broad adoption over the last year, and resulted in a large amount of feedback. We see this period to have been a critical maturation process for the Karpenter project, and has given us confidence that the changes in v1beta1 will be sufficient to promote after a shorter feedback period.
+
+## Migration
+
+**Tenet:** Customers will be able to migrate from v1alpha5 to v1beta1 in a single cluster using any Karpenter version from the time that we release v1beta1 up until we release v1.
+
+Kubernetes custom resources have built in support for API version compatibility. CRDs with multiple versions must define a “storage version”, which controls the data stored in etcd. Other versions are views onto this data and converted using [conversion webhooks](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion). However, there is a fundamental limitation that [all versions must be safely round-trippable through each other](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html)[.](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html) This means that it must be possible to define a function that converts a v1alpha5 Provisioner into a v1beta1 Provisioner and vise versa.
+
+Unfortunately, multiple proposed changes in v1beta1 are not round-trippable. Below, we propose deprecations of legacy fields in favor more modern mechanisms that have seen adoption in v1alpha5. These changes remove sharp edges that regularly cause users surprises and production pain.
+
+To workaround the limitation of round-trippability, we are proposing a rename of both the API group and Kind (`NodePool`) that the CRDs exist within. This allows both CRDs to exist alongside each other simultaneously and gives users a natural migration path to move through.
+
+### Migration Path
+
+#### Aggressive Scale-Down
+
+1. Spin up a new `v1beta1/NodePool` and `v1beta1/NodeClass` resource matching their `v1alpha5` counterparts
+2. Delete the `v1alpha5/Provisioner` and `v1alpha5/AWSNodeTemplate` with [cascading foreground deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#foreground-deletion) to delete the nodes and have Karpenter roll the nodes to `v1beta1`
+
+#### Slow Scale-Down
+
+1. Spin up a new `v1beta1/NodePool` and `v1beta1/NodeClass` resource matching their `v1alpha5` counterparts
+2. Set the `spec.limits` on your `v1alpha5/Provisioner` resource to `cpu=0` to stop provisioning
+3. Manually delete Nodes one-by-one to have Karpenter roll the nodes over to `v1beta1`
+
+## APIs
+
+To help clearly define where configuration should live within Karpenter’s API, we should define the logical boundary between each Kind in the project.
+
+1. `NodePool`
```

Comment:
> Starting with NodePool. I think it has a lot of abstraction leaks between App dev vs Cluster Operator. Ideally app devs can specify their pods -> node, maybe some taints and label for their nodes (pretty much all the parameters in spec). However app dev shouldn't be able to specify kubelet configurations since this is usually managed by cluster operator. I think if possible this can be carved out to something like NodeProviders which are associated to NodePool. Similar to HTTPRoutes to Gateway in GatewayAPI. I think this would significantly improve the flexibility of how cluster operators manage Provisioners/ NodePool since often times there are much more app dev teams to cluster operator teams and most cluster operators are currently creating provisioners per app team.

To further expand on this thought, I think there should be a separation between App Dev configuration and cluster operator. 
- An app dev entity that defines Node <-> pod compatibility on requirements, taints, label.
- A cluster operator entity that defines cloud-provider agnostic configuration for nodes (kubelet config), scheduling and deprovisioning decision making. Since these usually have deeper company wide polices such as SOX compliance, node lifetime policies, daemonset/ OS overhead, networking setup and such. Ideally an app dev doesn't even have to know this is "kubernetes"

Reply:
Here's how I'd solve that, **if needed**:
- don't put any behavior into Karpenter itself
- provide a contrib controller (or a configuration for `shell-operator`, or whatever) that allows someone to manage a subset of the settings for a NodePool

_or_

- a whole new controller that manages scheduling behavior and node outcomes, per namespace. Again, not Karpenter.

Reply:
To add to what @sftim said, there's tools like [Backstage](https://backstage.io/docs/features/kubernetes/) that effectively allow cluster admins to "wrap" Kubernetes with a simpler interface that they can hand off to their app teams. It's effectively templating the manifests but in a managed way. That might be a good solution here if there's a need to draw a line between what service teams should be owning and what cluster admins should be owning.

Reply:
interesting, let me look into that

Reply:
Out of curiosity, is there a reason why not to split it at the Karpenter level? Is this just adding too much to the API surface?

Reply:
Yeah, I think it's mainly trying to address too many concerns in one place. IMO, these should be layers of abstraction that users that need it can add as they please.